src/consume/mod.rs

```diff
@@ -0,0 +1,201 @@
+//! Types, traits, and functions necessary to consume messages using hedwig
+//!
+//! See the [`Consumer`] trait and [`consume`] function as entry points.
+
+use crate::ValidatedMessage;
+use async_trait::async_trait;
+use either::Either;
+use futures_util::stream;
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Message consumers ingest messages from a queue service and present them to the user application
+/// as a [`Stream`](futures_util::stream::Stream).
+///
+/// ## Acknowledging Messages
+/// Typically message services deliver messages with a particular delivery time window, during
+/// which this message won't be sent to other consumers. In AWS SQS this is called the [visibility
+/// timeout][AWS], and in GCP PubSub this is the [ack deadline][GCP].
+///
+/// If a message is successfully acknowledged within this time, it will be considered processed and
+/// not delivered to other consumers (and possibly deleted depending on the service's
+/// configuration). A message can conversely be negatively-acknowledged, to indicate e.g.
+/// processing has failed and the message should be delivered again to some consumer. This time
+/// window can also be modified for each message, to allow for longer or shorter message processing
+/// than the default configured time window.
+///
+/// Implementations of this trait do not ack/nack/modify messages themselves, and instead present
+/// this functionality to users with the [`AcknowledgeableMessage`] type. Message processors are
+/// responsible for handling message acknowledgement, including extensions for processing time as
+/// necessary.
+// If we had async drop, sending nacks on drop would be nice. Alas, rust isn't there yet
+///
+/// Bear in mind that message delivery and acknowledgement are all best-effort in distributed
+/// message services. An acknowledged or extended message may still be re-delivered for any number
+/// of reasons, and applications should be made resilient to such events.
+///
+/// [AWS]: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html
+/// [GCP]: https://cloud.google.com/pubsub/docs/subscriber
+pub trait Consumer {
+    /// The type of acknowledgement tokens produced by the underlying service implementation
+    type AckToken: AcknowledgeToken;
+    /// Errors encountered while streaming messages
+    type Error: std::error::Error + Send + Sync + 'static;
+    /// The stream returned by [`stream`]
+    type Stream: stream::Stream<
+        Item = Result<AcknowledgeableMessage<Self::AckToken, ValidatedMessage>, Self::Error>,
```

Comment:
I think this probably warrants a name change for `ValidatedMessage` because at this point the payload has not been validated (quite the opposite, in fact).

How about `Payload` or `MessagePayload` or something along those lines?

Reply:
I dunno, it _was_ validated on the writing side by some validator, with a schema that defines which one. I'm not against a name change though

Reply:
is it also validated on the consumer side later on?

Reply:
for protobufs the schema is checked; if the schema is as expected, then prost will decode it. So validation happens in the explicit schema check, and then whatever protobuf checks happen internally

Reply:
> it was validated on the writing side by some validator

Relying on this would introduce an assumption that all the clients are using the same version of the validator or schema or validate messages in the same way. As all the services are distinct and may be written in different languages (that's the point of hedwig!) there may be mismatches between the publishers and the consumers in terms of their expectations.

But I don't mind leaving this up to a later improvement of some sort.

Reply:
maybe it should just be called `SchemaBackedMessage` ?

Reply:
> that all the clients are using the same version of the validator or schema or validate messages in the same way

I get what you're saying, but this gets into a rabbit hole of having fields that define who is using what version of what implementation -- I don't think that those are usefully separate, I claim that all of those properties _should be_ encoded in the schema. If two validators, in any language, yield different results for the same schema, that should be a programming bug (which we can't reliably protect against).